### PR TITLE
Implement parser support for all select modifiers

### LINKED ISSUE #15318: Support all SELECT modifiers
`SELECT` in mysql supports a range of modifiers, such as `DISTINCT` and `SQL_CALC_FOUND_ROWS`.
At the moment, Vitess does not support:

```
[HIGH_PRIORITY]
[SQL_SMALL_RESULT]
[SQL_BIG_RESULT]
[SQL_BUFFER_RESULT]
```

It would be nice to have our parser support all of these. More info can be found [here](https://dev.mysql.com/doc/refman/8.0/en/select.html).

### PR DESCRIPTION
**Pull request by Utkar5hM**
## Description

Implemented support for the following select modifiers in the sqlparser:
```
[HIGH_PRIORITY]
[SQL_SMALL_RESULT]
[SQL_BIG_RESULT]
[SQL_BUFFER_RESULT]
```

Added simple test cases for each of the modifiers under `go/vt/sqlparser/parse_test.go` similarly to that of `sql_calc_found_rows`.

I'll update the semantics check for the following modifiers and `sql_calc_found_rows` in a separate PR as discussed on slack. 

## Related Issue(s)

Fixes #15318 

## Checklist

-   [ ] "Backport to:" labels have been added if this change should be back-ported to release branches
-   [ ] If this change is to be back-ported to previous releases, a justification is included in the PR description
-   [x] Tests were added or are not required
-   [ ] Did the new or modified tests pass consistently locally and on CI?
-   [x] Documentation was added or is not required

### CODE REVIEW

**Comment by dbussink** [comment]
@Utkar5hM You'll also need to update the other test fixtures. If you run them, it should tell you how you can copy over the new expectations.

**Comment by Utkar5hM** [comment]
- updated the test cases to support the select modifiers. 

- I've added `SQL_BUFFER_RESULT` under non reserved keywords list in `sql.y`.

- made a change to the func `testFile()` to make the newly generated `select_case.txt` be easily copyable in case it's the correct one. Previously, a error generating statement would still produce a  `error` output text even if the error is null and there is an actual output which would require a lot of manual work changing `ERROR` to `OUTPUT` and regenerating `select_case.txt`.

- However the particular test case at https://github.com/vitessio/vitess/blob/main/go/vt/sqlparser/testdata/select_cases.txt#L19706 which errored due to sql_big_result not being a modifier now passes.
```
INPUT
select sql_big_result distinct t1.a from t1,t2 order by t2.a;
END
ERROR
syntax error at position 31 near 'distinct'
END
```
now produces
```
OUTPUT
select distinct sql_big_result t1.a from t1, t2 order by t2.a asc
```
which I think should fail as it is still an incorrect query as can be seen below:
image

File: go/vt/sqlparser/sql.y [hunk 1]
```diff
@@ -8497,6 +8513,7 @@ non_reserved_keyword:
 | SNAPSHOT
 | SOME %prec ANY_SOME
 | SQL
+| SQL_BUFFER_RESULT
```
Comment (GuptaManan100) [inline go/vt/sqlparser/sql.y:8516]: `HIGH_PRIORITY`, `SQL_SMALL_RESULT` and `SQL_BIG_RESULT` need to be added to this list too.

**Comment by Utkar5hM** [comment]
I have updated the above. 

Addition of `SQL_BUFFER_RESULT` seems to have caused 3 shift/reduce conflicts making `static Code check` workflow to fail.  
```
1732: shift/reduce conflict (shift 1738(0), red'n 901(0)) on SQL_BUFFER_RESULT
state 1732
	select_options_opt:  select_options.    (901)
	select_options:  select_options.select_option 

	DISTINCT  shift 1734
	ALL  shift 1742
	DISTINCTROW  shift 1735
	SQL_NO_CACHE  shift 2367
	SQL_CACHE  shift 2368
	SQL_CALC_FOUND_ROWS  shift 1741
	SQL_SMALL_RESULT  shift 1739
	SQL_BIG_RESULT  shift 1740
	SQL_BUFFER_RESULT  shift 1738
	HIGH_PRIORITY  shift 1736
	STRAIGHT_JOIN  shift 1737
	.  reduce 901 (src line 4801)

	select_option  goto 2372
```

```
1727: shift/reduce conflict (shift 1738(0), red'n 900(0)) on SQL_BUFFER_RESULT
state 1727
	query_primary:  SELECT comment_opt.select_options_opt select_expression_list into_clause from_opt where_expression_opt group_by_opt having_opt named_windows_list_opt 
	query_primary:  SELECT comment_opt.select_options_opt select_expression_list from_opt where_expression_opt group_by_opt having_opt named_windows_list_opt 
	select_options_opt: .    (900)

	DISTINCT  shift 1734
	ALL  shift 1742
	DISTINCTROW  shift 1735
	SQL_NO_CACHE  shift 2367
	SQL_CACHE  shift 2368
	SQL_CALC_FOUND_ROWS  shift 1741
	SQL_SMALL_RESULT  shift 1739
	SQL_BIG_RESULT  shift 1740
	SQL_BUFFER_RESULT  shift 1738
	HIGH_PRIORITY  shift 1736
	STRAIGHT_JOIN  shift 1737
	.  reduce 900 (src line 4797)

	select_options  goto 1732
	select_options_opt  goto 1729
	select_option  goto 1733
```

```
1037: shift/reduce conflict (shift 1738(0), red'n 900(0)) on SQL_BUFFER_RESULT
state 1037
	query_expression:  SELECT comment_opt.cache_opt NEXT num_val for_from table_name 
	query_primary:  SELECT comment_opt.select_options_opt select_expression_list into_clause from_opt where_expression_opt group_by_opt having_opt named_windows_list_opt 
	query_primary:  SELECT comment_opt.select_options_opt select_expression_list from_opt where_expression_opt group_by_opt having_opt named_windows_list_opt 
	cache_opt: .    (887)
	select_options_opt: .    (900)

	DISTINCT  shift 1734
	ALL  shift 1742
	DISTINCTROW  shift 1735
	NEXT  reduce 887 (src line 4732)
	SQL_NO_CACHE  shift 1730
	SQL_CACHE  shift 1731
	SQL_CALC_FOUND_ROWS  shift 1741
	SQL_SMALL_RESULT  shift 1739
	SQL_BIG_RESULT  shift 1740
	SQL_BUFFER_RESULT  shift 1738
	HIGH_PRIORITY  shift 1736
	STRAIGHT_JOIN  shift 1737
	.  reduce 900 (src line 4797)

	cache_opt  goto 1728
	select_options  goto 1732
	select_options_opt  goto 1729
	select_option  goto 1733
```
I'm pretty new to yacc and been having a look into how to eliminate this conflicts. I might take some time understanding it.

**Comment by GuptaManan100** [comment]
I have fixed the shift-reduce conflicts for you.

**Review by GuptaManan100 — APPROVED** [review]

File: go/vt/sqlparser/testdata/select_cases.txt [hunk 2]
```diff
@@ -3080,7 +3080,7 @@ INPUT
 select sql_big_result score,count(*) from t1 group by score order by score desc;
 END
 OUTPUT
-select `sql_big_result` as score, count(*) from t1 group by score order by score desc
+select sql_big_result score, count(*) from t1 group by score order by score desc
```
Comment (systay) [inline go/vt/sqlparser/testdata/select_cases.txt:3083]: this does not look correct to me

File: go/vt/sqlparser/testdata/select_cases.txt [hunk 3]
```diff
@@ -10640,7 +10640,7 @@ INPUT
 select sql_big_result c,count(t) from t1 group by c order by c limit 10;
 END
 OUTPUT
-select `sql_big_result` as c, count(t) from t1 group by c order by c asc limit 10
+select sql_big_result c, count(t) from t1 group by c order by c asc limit 10
```
Comment (systay) [inline go/vt/sqlparser/testdata/select_cases.txt:10643]: also not correct

File: go/vt/sqlparser/testdata/select_cases.txt [hunk 4]
```diff
@@ -20852,7 +20852,7 @@ INPUT
 select SQL_BIG_RESULT a,count(b), sum(b), avg(b), std(b), min(b), max(b), bit_and(b), bit_or(b) from t1 group by a;
 END
 OUTPUT
-select `SQL_BIG_RESULT` as a, count(b), sum(b), avg(b), std(b), min(b), max(b), bit_and(b), bit_or(b) from t1 group by a
+select sql_big_result a, count(b), sum(b), avg(b), std(b), min(b), max(b), bit_and(b), bit_or(b) from t1 group by a
```
Comment (systay) [inline go/vt/sqlparser/testdata/select_cases.txt:20855]: not right

**Review by systay — CHANGES_REQUESTED** [review]
When writing something like:

```sql
select `SQL_BIG_RESULT` as a
```

we should not parse this as a SELECT modifier - this is a escaped column name.

File: go/vt/sqlparser/testdata/select_cases.txt (hunk 4 above)
Comment (systay) [reply go/vt/sqlparser/testdata/select_cases.txt:20855]: @GuptaManan100 explained it to me. I was reading it wrong...

**Review by systay — APPROVED** [review]
I was reading it wrong. This LGTM!

**Pull request merged**